### PR TITLE
Adding reference to Nuget package of LevelDB.Standard

### PR DESCRIPTION
**Pull request by eryeer**
Here is the reference change of LevelDB and the unit test of Persistence module that related to the change.

### CODE REVIEW

File: neo.UnitTests/TestUtils.cs [hunk 1]
```diff
@@ -108,5 +108,29 @@ public static Transaction CreateRandomHashTransaction()
 
             return newObj;
         }
+
+        public static void DeleteFile(string file)
+        {
```
Comment (shargon) [inline neo.UnitTests/TestUtils.cs]: Rename it to `DeleteFiles` or `DeleteFolder`

File: neo/Persistence/LevelDB/DbCache.cs [hunk 2]
```diff
@@ -18,19 +19,19 @@ public class DbCache<TKey, TValue> : DataCache<TKey, TValue>
         public DbCache(DB db, ReadOptions options, WriteBatch batch, byte prefix)
         {
             this.db = db;
-            this.options = options ?? ReadOptions.Default;
+            this.options = options ?? new ReadOptions();
```
Comment (shargon) [inline neo/Persistence/LevelDB/DbCache.cs]: this new is IDisposable, we should Dispose it in this class too

File: neo/Persistence/LevelDB/DbMetaDataCache.cs [hunk 3]
```diff
@@ -17,26 +17,25 @@ public DbMetaDataCache(DB db, ReadOptions options, WriteBatch batch, byte prefix
             : base(factory)
         {
             this.db = db;
-            this.options = options ?? ReadOptions.Default;
+            this.options = options ?? new ReadOptions();
```
Comment (shargon) [inline neo/Persistence/LevelDB/DbMetaDataCache.cs]: Same here, IDisposable

File: neo/neo.csproj [hunk 4]
```diff
@@ -23,6 +23,7 @@
   <ItemGroup>
     <PackageReference Include="Akka" Version="1.3.11" />
     <PackageReference Include="K4os.Compression.LZ4" Version="1.1.3" />
+    <PackageReference Include="LevelDB.Standard" Version="2.1.5" />
```
Comment (shargon) [inline neo/neo.csproj]: I am scared about this, i love nugets, but this is a very humble nugget, so how could we are sure that the owner don't cheat us?

File: neo.UnitTests/TestUtils.cs (hunk 1 above)
Comment (shargon) [reply neo.UnitTests/TestUtils.cs]: Is not used?

File: neo/neo.csproj (hunk 4 above)
Comment (eryeer) [reply neo/neo.csproj]: @erikzhang , please please take a consideration about this.

File: neo/Persistence/LevelDB/DbMetaDataCache.cs [hunk 5]
```diff
@@ -37,5 +38,14 @@ protected override void UpdateInternal(T item)
         {
             batch?.Put(new byte[] { prefix }, item.ToArray());
         }
+
+        public override void Dispose()
+        {
+            if (options != null)
+            {
+                MethodInfo method = options.GetType().GetMethod("FreeUnManagedObjects", BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (shargon) [inline neo/Persistence/LevelDB/DbMetaDataCache.cs]: it haven't got Dispose? this reflection is very slow

File: neo/Persistence/LevelDB/DbCache.cs [hunk 6]
```diff
@@ -53,5 +54,13 @@ protected override void UpdateInternal(TKey key, TValue value)
         {
             batch?.Put(SliceBuilder.Begin(prefix).Add(key).ToArray(), value.ToArray());
         }
+
+        public override void Dispose()
+        {
+            if (options != null) {
+                MethodInfo method = options.GetType().GetMethod("FreeUnManagedObjects", BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (shargon) [inline neo/Persistence/LevelDB/DbCache.cs]: it haven't got Dispose? this reflection is very slow

File: neo/Persistence/LevelDB/DbMetaDataCache.cs (hunk 5 above)
Comment (eryeer) [reply neo/Persistence/LevelDB/DbMetaDataCache.cs]: Oh yes, I made a mistake.

File: neo/neo.csproj (hunk 4 above)
Comment (lock9) [reply neo/neo.csproj]: @shargon  can't we "have our own"? Like "neo leveldb" in nuget?

File: neo.UnitTests/Persistence/UT_Store.cs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+        [TestMethod]
+        public void TestGetBlocks()
+        {
+            Snapshot snapshot = store.GetSnapshot();
+            TrimmedBlock block = new TrimmedBlock();
+            block.ConsensusData = new ConsensusData();
+            block.MerkleRoot = UInt256.Parse("0xa400ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff02");
+            block.PrevHash = UInt256.Parse("0xa400ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff00ff01");
+            block.Timestamp = new DateTime(1968, 06, 01, 0, 0, 0, DateTimeKind.Utc).ToTimestamp();
```
Comment (lock9) [inline neo.UnitTests/Persistence/UT_Store.cs]: Hello,
This date is not "adequate". It is before the "big bang" (01/01/1970). Could you use a more "reliable" date? I don't know the effects of this usage. Probably an overflow?

**Comment by lock9** [comment]
I have tons of issues with the current leveldb. It takes more than 1 hour for a new developer to compile and make it work properly. I think this PR is great, I wish we could add this to NEO 2, but I don't think we have the time for that.

**Comment by shargon** [comment]
I love the PR too, in fact i think that this is needed, but at same time i am scared about it. In the native code is so easy to hide a backdoor ... at least, we should release neo version with our own native leveldb.dll

File: neo.UnitTests/Persistence/UT_Store.cs (hunk 7 above)
Comment (eryeer) [reply neo.UnitTests/Persistence/UT_Store.cs]: OK，I will update it. Perhaps there is a risk when the date is before 1970.

File: neo/neo.csproj (hunk 4 above)
Comment (erikzhang) [reply neo/neo.csproj]: This is a problem.

We also have a leveldb package on nuget, but it does not currently support .NET Core.

https://www.nuget.org/packages/LevelDB.Net/

Maybe we should modify this package and use it?

Comment (shargon) [reply neo/neo.csproj]: I prefer to fork it and modify it. Then will be owned by neo.

**Comment by devhawk** [comment]
@lock9 Can you share the instructions for compiling leveldb for use with NEO?

**Comment by devhawk** [comment]
Reminder that C# can hide back doors too. Both Akka and K4os.Compression.LZ4 use unsafe code. Whatever process and policy used to monitor native dependencies like leveldb should also be used to monitor managed dependencies as well.

**Comment by igormcoelho** [comment]
I love the unit tests on this PR, and I'm neutral regarding leveldb. If it's system managed, or nuget managed, it doesn't add extra security in my opinion.

**Comment by devhawk** [comment]
Package management is clearly *NOT* a security system: https://haacked.com/archive/2019/06/11/package-mwanager-security/

File: neo/neo.csproj (hunk 4 above)
Comment (devhawk) [reply neo/neo.csproj]: I was able to compile leveldb from google's repo and deploy it to work with neo-cli locally. So wouldn't it make more sense to start with the google's source than any other fork of the leveldb code?

**Comment by Tommo-L** [comment]
Hi everyone, could we fork it, check and modify it, then release a neo version finally. It's easy for us to practice by this way. And it's more simpler for users by using nuget , they don't need to install the leveldb by themselves.  😂

**Comment by erikzhang** [comment]
We are going to use RocksDB, I think. #966 😂

**Comment by eryeer** [comment]
I will close this pr, since this change is replaced by RocksDB change.

**Comment by shargon** [comment]
This should be migrated to a plugin when Istorage is ready

**Comment by eryeer** [comment]
> This should be migrated to a plugin when Istorage is ready

Sure, I can reopen it or create a new PR then.

**Comment by shargon** [comment]
I am waiting with RocksDB opened as remember, when IStorage is ready, we can move it to `neo-plugins`

**Pull request closed without merging**